sematic/ui/src/components/ScrollingLogView.tsx

```diff
@@ -0,0 +1,218 @@
+import { useEffect, useState, useCallback, useMemo } from "react";
+import { Box, Button, TextField } from "@mui/material";
+import InfiniteScroll from "react-infinite-scroll-component";
+import Loading from "./Loading";
+
+export type MoreLinesCallback = (
+  source: string,
+  usedFilter: string,
+  lines: string[],
+  cursor: string | null,
+  noLinesReason: string | null
+) => void;
+export type GetLines = (
+  source: string,
+  cursor: string | null,
+  filterString: string,
+  moreLinesCallback: MoreLinesCallback
+) => void;
+
+const DEFAULT_NO_LINES_REASON = "No more matching lines";
+
+export default function ScrollingLogView(props: {
+  getLines: GetLines;
+  logSource: string;
+}) {
+  const { getLines, logSource } = props;
+
+  const [hasMore, setHasMore] = useState(true);
+  const [fastForwarding, setFastForwarding] = useState(false);
+  const [currentNoLinesReason, setNoLinesReason] = useState<string | null>(
+    DEFAULT_NO_LINES_REASON
+  );
+  const [loadingMessage, setLoadingMessage] = useState<string>("Loading...");
+
+  const [lineState, setLineState] = useState<{
```

Comment:
All the things relating to the current lines and where they came from should be atomically modified